### PR TITLE
Update inbound peer eviction and add advanced test coverage - Closes #4280

### LINKED ISSUE #4280: Advanced test coverage related to inbound eviction
### Description

More tests are needed to verify that inbound peer eviction is working under unusual scenarios.
Some things to check (note that this is not an exhaustive list):

- Verify that many bad peers connecting to a node quickly will not cause all good peers to be evicted.
- Verify that old version peers are not given more protection than new version peers.

Related issue: https://github.com/LiskHQ/lisk-sdk/issues/4229

### Motivation

Inbound peer eviction/protection is a complex feature which requires more thorough testing.

### Acceptance Criteria

- The test cases cover advanced or malicious scenarios like the ones described above.
- The goal of this issue is not to add test cases to verify obvious functionality. The tests should aim to uncover flaws. They need to focus on edge cases.

### PR DESCRIPTION
**Pull request by mitsujutsu**
### What was the problem?

- Some categories in inbound eviction process were dependent on each other causing minor change in eviction candidate results.

- There were no advanced unit tests for inbound eviction process.

### How did I solve it?

Decoupled `netgroup`, `latency` and `responseRate` peer eviction processing. 
Added required tests simulating different inbound peer scenarios.

The expected number of eviction candidates for each test are based on the following estimates derived from default protection ratios assuming 100 peers:  

`netgroup`: 4 peers, `latency`: 7 peers, `usefulness`: 7 peers, `longevity`: 41 peers

### How to manually test it?

npm run test

### Review checklist

- [ ] The PR resolves #4280 
- [ ] All new code is covered with unit tests
- [ ] Relevant integration / functional tests are added
- [ ] Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
- [ ] Documentation has been added/updated

### CODE REVIEW

File: elements/lisk-p2p/src/peer_pool.ts [hunk 1]
```diff
@@ -80,26 +80,30 @@ interface FilterPeersOptions {
 	readonly asc: boolean;
 }
 
-const filterPeersByCategory = (
+// Returns an array of peers to be protected
+export const filterPeersByCategory = (
```
Comment (ishantiw) [inline elements/lisk-p2p/src/peer_pool.ts]: Can we make this function to accept argument `sort`, that takes `ASC` and `DESC` enum instead of because when we give `asc` as `false` then it should not mean that we have to return descending. Also, in the framework api module has a sort field that accepts `ASC` and `DESC`.

File: elements/lisk-p2p/test/unit/peer_pool.ts [hunk 2]
```diff
@@ -73,6 +79,216 @@ describe('peerPool', () => {
...
+		describe('when node with netgroup protection disabled has 100 inbound peers', () => {
+			beforeEach(() => {
+				(peerPool as any)._peerPoolConfig.netgroupProtectionRatio = 0;
+				sandbox.stub(peerPool, 'getPeers').returns(originalPeers as Peer[]);
+			});
+
+			it('should return expected amount of eviction candidates', async () => {
+				const selectedPeersForEviction = (peerPool as any)._selectPeersForEviction();
+
+				expect(selectedPeersForEviction.length).to.eql(44);
```
Comment (ishantiw) [inline elements/lisk-p2p/test/unit/peer_pool.ts]: how we are expecting that there will be 44 peers? I see in most of the places we are using a specific number and we don't know why we are expecting that number. If it's possible can we break it down like `peersWithPoorLatency` and we can use `expect(selectedPeersForEviction.length).to.eql(peersWithPoorLatency.length)`

Comment (mitsujutsu) [reply elements/lisk-p2p/test/unit/peer_pool.ts]: If we compare to a `peersWithPoorLatency`, how do we define the expected threshold for `poorLatency`? The specific numbers can be calculated using the `DEFAULT` values. For example for this test, we have `netgroup` protection disabled. Using default protection ratios, the breakdown for 100 peers is: 4 by `netgroup`, 7 by `latency`, 7 by `usefulness `and 41 by `longevity` (simply multiply the ratio with 100). Since `netgroup` is disabled, that leaves `7 (latency) + 7(usefulness) + 41 (longevity) = 55. 100 - 55 = 45.` The 1 difference is a rounding difference, I opted to round down.

File: elements/lisk-p2p/src/peer_pool.ts [hunk 3]
```diff
@@ -603,53 +607,58 @@ export class PeerPool extends EventEmitter {
...
 					percentage: this._peerPoolConfig.longevityProtectionRatio,
-					asc: true,
+					asc: false,
 			  })
-			: filteredPeersByResponseRate;
+			: [];
 
-		return filteredPeersByConnectTime;
+		return unprotectedPeers.filter(
+			peer => !protectedPeersByConnectTime.includes(peer),
```
Comment (jondubois) [inline elements/lisk-p2p/src/peer_pool.ts]: It's better to avoid passing object references to `Array.includes()` in case the peer object is a clone then then references will not match (even though both objects represent the same thing).

Also, using `includes()` inside a loop could be expensive if there are a lot of peers.

You could add an additional step before this and create a `new Set` containing of all the projected peer IDs for the category. Then inside the `filter` loop, you just do `set.has(...)`; set lookup is very fast regardless of how many items are in the set.

This should also be done in other cases above where `include` is used.

File: elements/lisk-p2p/src/peer_pool.ts [hunk 4]
```diff
@@ -603,53 +607,58 @@ export class PeerPool extends EventEmitter {
...
-			return filteredPeersByResponseRate;
-		}
+			: [];
+
+		const uniqueProtectedPeers = [
+			...new Set([
+				...protectedPeersByNetgroup,
+				...protectedPeersByLatency,
+				...protectedPeersByResponseRate,
+			]),
```
Comment (jondubois) [inline elements/lisk-p2p/src/peer_pool.ts]: Why put the Set into an array? The Set is much faster for lookup especially if there are a lot of peers.

In the worst case, `Array.includes()` iterates over every element whereas `Set.has()` hashes directly to the key. Also, we should store the peer ids instead of the whole `Peer` instances or else it can lead to mismatched references if one object is a clone (which is a common thing that pure functions do); it's safer to always use primitive types for lookups.

File: elements/lisk-p2p/src/peer_pool.ts (hunk 1 above)
Comment (jondubois) [reply elements/lisk-p2p/src/peer_pool.ts]: Maybe instead of `asc` or `desc` it can just be an enum `highest` or `lowest` because it's not clear from outside how the sorting relates to the selection (at first, it's not clear if the function protects the first `x%` or the last `x%`? If the developer didn't read the implementation of the function, they won't know whether they should use asc or desc so `highest` and `lowest` is more clear).

File: elements/lisk-p2p/src/peer_pool.ts [hunk 5]
```diff
@@ -603,53 +607,58 @@ export class PeerPool extends EventEmitter {
...
-		const filteredPeersByConnectTime = this._peerPoolConfig
+		// Protect *the remaining half* of peers by longevity, precludes attacks that start later.
+		const protectedPeersByConnectTime = this._peerPoolConfig
 			.longevityProtectionRatio
-			? filterPeersByCategory(filteredPeersByResponseRate, {
+			? filterPeersByCategory(unprotectedPeers, {
 					category: PROTECTION_CATEGORY.CONNECT_TIME,
 					percentage: this._peerPoolConfig.longevityProtectionRatio,
-					asc: true,
+					asc: false,
```
Comment (jondubois) [inline elements/lisk-p2p/src/peer_pool.ts]: Doesn't this mean that those with the shortest connected time are more likely to be protected? We need to protect the oldest ones with the longest connected time.

File: elements/lisk-p2p/test/unit/peer_pool.ts [hunk 6]
```diff
@@ -73,6 +79,216 @@ describe('peerPool', () => {
...
+			const filteredPeers = filterPeersByCategory(originalPeers as any, {
+				category: PROTECTION_CATEGORY.RESPONSE_RATE,
+				percentage: 0.2,
+				asc: true,
+			});
+
+			expect(filteredPeers.filter(p => p.responseRate === 1).length).to.eql(2);
+		});
+
+		it('should protect peers with lowest connectTime value when sorted by descending', async () => {
```
Comment (jondubois) [inline elements/lisk-p2p/test/unit/peer_pool.ts]: Highest connectTime is a more important to test case.

File: elements/lisk-p2p/src/peer_pool.ts (hunk 5 above)
Comment (mitsujutsu) [reply elements/lisk-p2p/src/peer_pool.ts]: No, it means those who connected at the smallest `Date` number e.g. longest connected time.

File: elements/lisk-p2p/test/unit/peer_pool.ts (hunk 6 above)
Comment (mitsujutsu) [reply elements/lisk-p2p/test/unit/peer_pool.ts]: `connectTime` is the exact time of connect, not `duration` of connection time.

Comment (jondubois) [reply elements/lisk-p2p/test/unit/peer_pool.ts]: Ah nevermind then  ;p

File: elements/lisk-p2p/test/unit/peer_pool.ts (hunk 2 above)
Comment (ishantiw) [reply elements/lisk-p2p/test/unit/peer_pool.ts]: I mean if it's possible then can we actually make few variables (`noOfPeersbyNetgroup`, `noOfPeersbyLatency`, .etc) and then sum them up before using it in expect? Maybe just for a test so that its clear what's going on if I'm new and I read the tests directly.

Comment (mitsujutsu) [reply elements/lisk-p2p/test/unit/peer_pool.ts]: It's a good point. I will add a comment detailing the above calculation before the `expect` to make it clear.
